sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,539 @@
+provided by the user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void LibDeviceFunc();
+...
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([]() { LibDeviceFunc(); });
+});
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void LibDeviceFunc() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special compiler option
+(like `-fsycl-add-targets`).
+
+The main purpose of this feature is to provide a mechanism which allows to
+link device code dynamically at runtime, such as in the scenarios above.
+
+## Requirements:
+User's device code that consists of some device API (`SYCL_EXTERNAL` functions),
+is compiled into some form and it is not linked statically with device code of
+application. It can be a shared library that contains some device code or a
+separate device image supplied with properties attached. This code is linked
+dynamically at run time with device code of a user's application in order to
+resolve dependencies.
+For this combination the following statements must be true:
+
+The presented dynamic device code linkage mechanism must:
+
+- Allow to represent the actual dynamically linked code as a device binary image
+  which can be:
+  - Embedded into a host shared object by standard SYCL compiler driver
+    invocation
+  - Embedded into a host binary or shared object using manual invocations of
+    SYCL tools such as `clang-offload-wrapper` and linker
+  - Loaded into memory via special API
+- Must not assume the actual format of the device code - e.g. that it is SPIR-V
+  or native device binary
+- Provide automatic runtime resolution of `SYCL_EXTERNAL` function references
+  within the SYCL app to their definitions (if found) within any suitable
+  dynamically linked device binary image
+- Support pointers to `SYCL_EXTERNAL` functions across the dynamic linkage
+  boundaries within the device code - taking a pointer, call through a pointer.
+- Specific code changes are not required, i.e. the mechanism of linking works
+  as close as possible to host shared libraries.
+
+## Design
+The overall idea:
+
+- Each device image is supplied with a list of imported and exported symbol
+  names through device image properties mechanism
+- Before JIT-compiling a device image DPC++ RT will check if device image has a
+  list of imported symbols and if it has, then RT will search for device images
+  which define required symbols using lists of exported symbols.
+  - Besides symbol names, additional attributes are taken into account (like
+    device image format: SPIR-V or device asm)
+  - No logical binding between host module and export/import lists, i.e.
+    resolution is performed w/o regard to containing host modules
+- Actual linking is performed by underlying backend (OpenCL/L0/etc.)
+  - Underlying backend is the backend used by the SYCL RT to perform JIT
```

Comment:
> perform JIT compilation of the program with symbols that need dynamic resolution.

What about AOT mode, how is that supported?

Reply:
The design doesn't assume format of device images being linked. So, AOT can be supported for some BE if there is support for native binaries linking.

Reply:
OK, then why this is talking about JIT specifically. AOT binaries can participate in the linking too, I expect.

Reply:
You are right, I've tried to remove binding to JIT.